File: C++/Sorting/dutch_national_flag_algorithm.cpp
#include<bits/stdc++.h>
using namespace std;
void sort012(vector<int> &nums, int n) 
{
    int l = 0, m = 0, h = n-1;
    while(m <= h)
    {
        if(nums[m] == 0)
        {
            swap(nums[l], nums[m]);
            l++;
            m++;
        }
        else if(nums[m] == 1)
        {
            m++;
        }
        else
        {
            swap(nums[h], nums[m]);
            h--;
        }
    }
}

int main()
{
    int n;
    cin>>n;
    vector<int> nums(n);
    for(int i = 0; i < n; i++)
    {
        cin>>nums[i];
    }
    sort012(nums, n);
    for(int i = 0; i < n; i++)
    {
        cout<<nums[i]<<" ";
    }
    return 0;
}

File: Python/Sorting/dutch_national_flag_algorithm.py
#Time Complexity - O(N)
#Space Complexity - O(1)

def sort012(arr, n):

    #start left pointer from 0
    #start right pointer from n-1
    
    left = 0
    right = n-1
    i=0

    # If the ith element is 0 then we will swap the element with the low range.
    # Similarly, if the ith element is 1 then keep it as it is.
    # If the element is 2 then we swap it with an element in the high range.

    while i<=right:
        if arr[i] == 0:
            arr[left], arr[i] = arr[i], arr[left]
            left+=1
        elif arr[i] == 2:
            arr[right], arr[i] = arr[i], arr[right]
            right-=1
            i-=1
            
        i+=1

print('Enter the number of elements in array:')
n = int(input())

print('Enter the array values (in space separated format):')
arr = [int(i) for i in input().split(' ')]

sort012(arr, n)

print('Sorted array is:')
print(' '.join([str(i) for i in arr]))
                